### PR TITLE
Add Undo/Redo Functions to editor

### PR DESCRIPTION
**Pull request by rudrathedev**

### CODE REVIEW

File: codeview/src/main/java/com/amrdeveloper/codeview/CodeView.java [hunk 1]
```diff
@@ -559,6 +561,26 @@ public void setLineNumberTextSize(float size) {
     public void setMatchingHighlightColor(int color) {
         matchingColor = color;
     }
+    
+    public void undo(){
+    mUndoRedoManager.undo();
+    }
+    
+    public void redo(){
+    mUndoRedoManager.undo();
```
Comment (AmrDeveloper) [inline codeview/src/main/java/com/amrdeveloper/codeview/CodeView.java:570]: Undo and redo both call `mUndoRedoManager.undo();`

File: codeview/src/main/java/com/amrdeveloper/codeview/UndoRedoManager.java [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+            editor.putInt(pre + ".start", ei.mmStart);
+            editor.putString(pre + ".before", ei.mmBefore.toString());
+            editor.putString(pre + ".after", ei.mmAfter.toString());
+
+            i++;
+        }
+    }
+
+    public boolean restorePersistentState(SharedPreferences sp, String prefix)
+            throws IllegalStateException {
```
Comment (AmrDeveloper) [inline codeview/src/main/java/com/amrdeveloper/codeview/UndoRedoManager.java:112]: Why we need `throws IllegalStateException`? btw `doRestorePersistentState` didn't throw any exception

File: codeview/src/main/java/com/amrdeveloper/codeview/UndoRedoManager.java [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+            }
+            mEditHistory.add(new EditItem(start, before, after));
+        }
+
+        mEditHistory.mmPosition = sp.getInt(prefix + ".position", -1);
+        if (mEditHistory.mmPosition == -1) {
+            return false;
+        }
+
+        return true;
```
Comment (AmrDeveloper) [inline codeview/src/main/java/com/amrdeveloper/codeview/UndoRedoManager.java:159]: This should can simplify to 

```java
return mEditHistory.mmPosition != -1;
```

File: codeview/src/main/java/com/amrdeveloper/codeview/UndoRedoManager.java [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+    }
+
+    private final class EditHistory {
+      
+        private int mmPosition = 0;
```
Comment (AmrDeveloper) [inline codeview/src/main/java/com/amrdeveloper/codeview/UndoRedoManager.java:164]: Please change mm to single m

**Comment by AmrDeveloper** [comment]
Hello, @rudrathedev,

Thanks for your effort,

This code is good for Undo and Redo as we talk about the issue we need more general commands support and we need to design it to be customizable and easy,

So can you move your class UndoRedoManager to the app as an external helper class until we design our new commands support

After finishing it please squash your commits into a single one with commit, you can do it from CLI or Android Studio IDE

Thanks Again,
Amr Hesham

File: codeview/src/main/java/com/amrdeveloper/codeview/CodeView.java [hunk 5]
```diff
@@ -559,6 +561,26 @@ public void setLineNumberTextSize(float size) {
...
+    }
+    
+    public boolean canUndo(){
+    return mUndoRedoManager.canUndo();
```
Comment (AmrDeveloper) [inline codeview/src/main/java/com/amrdeveloper/codeview/CodeView.java:574]: Please make sure that your code is formatted

**Review by AmrDeveloper — CHANGES_REQUESTED** [review]

**Comment by AmrDeveloper** [comment]
Hello @rudrathedev any updates?

**Comment by m-anshuman2166** [comment]
@AmrDeveloper we should close this since it is already implemented by you

**Comment by AmrDeveloper** [comment]
You are right@m-anshuman2166, thank you

**Pull request closed without merging**